src/lib/snarky_field_extensions/field_extensions.ml

```diff
@@ -0,0 +1,222 @@
+module Make (F : Intf.Basic) = struct
+  open F.Impl
+  open Let_syntax
+  open F
+
+  let typ = F.typ
+  let constant = F.constant
+  let scale = F.scale
+  let assert_r1cs = F.assert_r1cs
+  let ( + ) = F.(+)
+  let ( - ) = F.(-)
+  let negate = F.negate
+
+  let zero = constant Unchecked.zero
+  let one = constant Unchecked.one
+
+  let assert_square =
+    match assert_square with
+    | `Custom f -> f
+    | `Define ->
+      fun a a2 -> assert_r1cs a a a2
+
+  let ( * ) =
+    match ( * ) with
+    | `Custom f -> f
+    | `Define ->
+      fun x y ->
+        let%bind res =
+          provide_witness typ
+            As_prover.(map2 (read typ x) (read typ y) ~f:Unchecked.( * ))
+        in
+        let%map () = assert_r1cs x y res in
+        res
+
+  let square =
+    match square with
+    | `Custom f -> f
+    | `Define ->
+      fun x ->
+        let%bind res =
+          provide_witness typ
+            As_prover.(map (read typ x) ~f:Unchecked.square)
+        in
+        let%map () = assert_square x res in
+        res
+
+  let inv =
```

Comment:
I think this should be called `inv_exn`. It's not unsafe in the same sense as I've been using unsafe elsewhere in the code, which is basically it can return a very large set of things in the unsafe case.